pkg/ccl/testccl/workload/schemachange/schemachange_test: TestWorkload failed

pkg/ccl/testccl/workload/schemachange/schemachange_test.TestWorkload [failed](https://mesolite.cluster.engflow.com/invocations/default/337ee386-92b1-4fd1-a64a-6536c2b40b26?testReportRun=1&testReportShard=1&testReportAttempt=1#targets-Ly9wa2cvY2NsL3Rlc3RjY2wvd29ya2xvYWQvc2NoZW1hY2hhbmdlOnNjaGVtYWNoYW5nZV90ZXN0) on master @ [fdf1c66c3a933edc3c462bcf893a562a7b142ce8](https://github.com/cockroachdb/cockroach/commits/fdf1c66c3a933edc3c462bcf893a562a7b142ce8):


```
   }
  ],
  "previousStatements": [
   "ALTER TABLE public.table_w0_70 ALTER PRIMARY KEY USING COLUMNS (col70_w0_78)"
  ]
 }
}
    schema_change_external_test.go:165: 
        	Error Trace:	pkg/ccl/testccl/workload/schemachange/schema_change_external_test.go:165
        	Error:      	Received unexpected error:
        	            	***UNEXPECTED COMMIT ERROR; Received an unexpected commit error: ERROR: transaction committed but schema change aborted with error: (XX000): executing declarative schema change PostCommitPhase stage 7 of 15 with 6 ValidationType ops (rollback=false) for ALTER TABLE: []*scop.ValidateIndex: [0xc0153a15e0 0xc0153a1780 0xc0153a1868 0xc0153a1958 0xc0153a1a48 0xc0153a1b38]: validation of non-unique index crdb_internal_index_11_name_placeholder failed: expected 5 rows, found 4 (SQLSTATE XX000)
        	            	(1) forced error mark
        	            	  | "fatal error when running txn"
        	            	  | github.com/cockroachdb/errors/withstack/*withstack.withStack::
        	            	Wraps: (2)
        	            	Wraps: (3) attached stack trace
        	            	  -- stack trace:
        	            	  | github.com/cockroachdb/cockroach/pkg/workload/schemachange.(*schemaChangeWorker).run
        	            	  | 	pkg/workload/schemachange/schemachange.go:771
        	            	  | pkg/ccl/testccl/workload/schemachange/schemachange_test_test.TestWorkload.TestWorkload.func5.func8
        	            	  | 	pkg/ccl/testccl/workload/schemachange/schema_change_external_test.go:154
        	            	  | golang.org/x/sync/errgroup.(*Group).Go.func1
        	            	  | 	external/org_golang_x_sync/errgroup/errgroup.go:93
        	            	  | runtime.goexit
        	            	  | 	src/runtime/asm_amd64.s:1693
        	            	Wraps: (4) ***UNEXPECTED COMMIT ERROR; Received an unexpected commit error
        	            	Wraps: (5) ERROR: transaction committed but schema change aborted with error: (XX000): executing declarative schema change PostCommitPhase stage 7 of 15 with 6 ValidationType ops (rollback=false) for ALTER TABLE: []*scop.ValidateIndex: [0xc0153a15e0 0xc0153a1780 0xc0153a1868 0xc0153a1958 0xc0153a1a48 0xc0153a1b38]: validation of non-unique index crdb_internal_index_11_name_placeholder failed: expected 5 rows, found 4 (SQLSTATE XX000)
        	            	Error types: (1) *markers.withMark (2) *schemachange.ErrorState (3) *withstack.withStack (4) *errutil.withPrefix (5) *pgconn.PgError
        	Test:       	TestWorkload
{
 "message": "Inspect Job Errors",
 "jobs": [
  {
   "jobId": 1163566583387881473,
   "status": "no errors reported"
  }
 ]
}
Schema Workload Stats
Total Schema Statements Executed = 50
Total Schema Statements Succeeded = 50
Total Schema Statement Expected Failures = 0
Total Transactions Committed = 35
Total Transactions Rolled Back = 42
Total Transactions Executed = 77
    schema_change_external_test.go:118: backup, tracing data, and system table dumps in /mnt/engflow/worker/work/0/exec/bazel-out/k8-fastbuild/testlogs/pkg/ccl/testccl/workload/schemachange/schemachange_test/test.outputs/logTestWorkload2973048932
    schema_change_external_test.go:127: running DROP with use_declarative_schema_changer = on
    panic.go:615: -- test log scope end --
test logs left over in: outputs.zip/logTestWorkload2973048932
--- FAIL: TestWorkload (39.60s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>run=1</code>
 - <code>shard=1</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #167005 pkg/ccl/testccl/workload/schemachange/schemachange_test: TestWorkload failed [C-test-failure O-robot P-2 T-sql-foundations branch-release-24.3 target-release-24.3.31]
</p>
</details>
/cc @cockroachdb/sql-foundations
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestWorkload.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-62456

pkg/ccl/testccl/workload/schemachange/schemachange_test.TestWorkload [failed](https://mesolite.cluster.engflow.com/invocations/default/5ad489e7-c657-4cbb-bef8-1cb59fe0ebd4?testReportRun=4&testReportShard=1&testReportAttempt=1#targets-Ly9wa2cvY2NsL3Rlc3RjY2wvd29ya2xvYWQvc2NoZW1hY2hhbmdlOnNjaGVtYWNoYW5nZV90ZXN0) on master @ [8b6d3d2a14f67c0867d9f492fde1fad929538a51](https://github.com/cockroachdb/cockroach/commits/8b6d3d2a14f67c0867d9f492fde1fad929538a51):


```
  ],
  "previousStatements": [
   "ALTER TYPE public.enum_w1_20 DROP VALUE 'va'"
  ]
 }
}
    schema_change_external_test.go:183: 
        	Error Trace:	pkg/ccl/testccl/workload/schemachange/schema_change_external_test.go:183
        	            				pkg/ccl/testccl/workload/schemachange/schema_change_external_test.go:41
        	Error:      	Received unexpected error:
        	            	***UNEXPECTED COMMIT ERROR; Received an unexpected commit error: ERROR: transaction committed but schema change aborted with error: (2BP01): could not remove enum value "va" as it is being used in a routine "udf_w1_146" (SQLSTATE 2BP01)
        	            	(1) forced error mark
        	            	  | "fatal error when running txn"
        	            	  | github.com/cockroachdb/errors/withstack/*withstack.withStack::
        	            	Wraps: (2)
        	            	Wraps: (3) attached stack trace
        	            	  -- stack trace:
        	            	  | github.com/cockroachdb/cockroach/pkg/workload/schemachange.(*schemaChangeWorker).run
        	            	  | 	pkg/workload/schemachange/schemachange.go:771
        	            	  | pkg/ccl/testccl/workload/schemachange/schemachange_test_test.runSchemaChangeWorkload.runSchemaChangeWorkload.func5.func7
        	            	  | 	pkg/ccl/testccl/workload/schemachange/schema_change_external_test.go:172
        	            	  | golang.org/x/sync/errgroup.(*Group).Go.func1
        	            	  | 	external/org_golang_x_sync/errgroup/errgroup.go:93
        	            	  | runtime.goexit
        	            	  | 	src/runtime/asm_amd64.s:1771
        	            	Wraps: (4) ***UNEXPECTED COMMIT ERROR; Received an unexpected commit error
        	            	Wraps: (5) ERROR: transaction committed but schema change aborted with error: (2BP01): could not remove enum value "va" as it is being used in a routine "udf_w1_146" (SQLSTATE 2BP01)
        	            	Error types: (1) *markers.withMark (2) *schemachange.ErrorState (3) *withstack.withStack (4) *errutil.withPrefix (5) *pgconn.PgError
        	Test:       	TestWorkload
{
 "message": "Inspect Job Errors",
 "jobs": [
  {
   "jobId": 1167632325602344961,
   "status": "no errors reported"
  }
 ]
}
Schema Workload Stats
Total Schema Statements Executed = 97
Total Schema Statements Succeeded = 97
Total Schema Statement Expected Failures = 0
Total Transactions Committed = 61
Total Transactions Rolled Back = 66
Total Transactions Executed = 127
    schema_change_external_test.go:136: backup, tracing data, and system table dumps in /mnt/engflow/worker/work/0/exec/bazel-out/k8-fastbuild/testlogs/pkg/ccl/testccl/workload/schemachange/schemachange_test/shard_1_of_2_run_4_of_10/test.outputs/logTestWorkload1214969766
    schema_change_external_test.go:145: running DROP with use_declarative_schema_changer = on
    panic.go:694: -- test log scope end --
test logs left over in: outputs.zip/logTestWorkload1214969766
--- FAIL: TestWorkload (63.87s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>run=4</code>
 - <code>shard=1</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #168378 pkg/ccl/testccl/workload/schemachange/schemachange_test: TestWorkload failed [C-test-failure O-robot T-sql-foundations branch-release-24.3.31-rc release-blocker]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestWorkload.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

--- a/pkg/workload/schemachange/operation_generator.go
+++ b/pkg/workload/schemachange/operation_generator.go
@@ -5919,25 +5919,43 @@ type triggerInfo struct {
 	triggerName string
 }
 
-// collectEnumMembers returns all enum members in the current database. Each
-// result row is a map with keys "name" (qualified enum type name), "value"
-// (quoted member literal), and "non_public" (bool indicating a DROPPING member).
+// collectEnumMembers returns all public enum members in the current database.
+// Each result row is a map with keys "name" (qualified enum type name), "value"
+// (quoted member literal), and "non_public" (always false).
+//
+// This uses enum_range(NULL::type) which goes through the lease-based type
+// resolution and filters out read-only members. This ensures the results are
+// consistent with what DML statements see, avoiding TOCTOU races where the
+// previous system.descriptor-based query could see a member as public while a
+// concurrent schema change has already marked it as READ_ONLY in the leased
+// descriptor.
 func (og *operationGenerator) collectEnumMembers(
 	ctx context.Context, tx pgx.Tx,
 ) ([]map[string]any, error) {
-	q := With([]CTE{
-		{"descriptors", descJSONQuery},
-		{"enums", enumDescsQuery},
-		{"enum_members", enumMemberDescsQuery},
-	}, `SELECT
-			quote_ident(schema_id::REGNAMESPACE::TEXT) || '.' || quote_ident(name) AS name,
-			quote_literal(member->>'logicalRepresentation') AS value,
-			(
-				COALESCE(member->>'direction' = 'REMOVE', false) OR
-				COALESCE(member->>'capability' = 'READ_ONLY', false)
-			) AS non_public
-		FROM enum_members
+	// First, collect all user-defined enum type names via pg_type (lease-based).
+	enumTypeNames, err := Collect(ctx, og, tx, pgx.RowTo[string], `
+		SELECT quote_ident(n.nspname) || '.' || quote_ident(t.typname)
+		FROM pg_catalog.pg_type t
+		JOIN pg_catalog.pg_namespace n ON t.typnamespace = n.oid
+		WHERE t.typtype = 'e'
+		  AND n.nspname NOT IN ('information_schema', 'pg_catalog', 'pg_extension', 'crdb_internal')
 	`)
+	if err != nil {
+		return nil, err
+	}
+	if len(enumTypeNames) == 0 {
+		return nil, nil
+	}
+
+	// For each enum type, use enum_range to get only public members. Build a
+	// UNION ALL query so we make a single round-trip.
+	var parts []string
+	for _, name := range enumTypeNames {
+		parts = append(parts, fmt.Sprintf(
+			`SELECT '%[1]s' AS name, quote_literal(v) AS value, false AS non_public
+			FROM unnest(enum_range(NULL::%[1]s)) v`, name))
+	}
+	q := strings.Join(parts, " UNION ALL ")
 	return Collect(ctx, og, tx, pgx.RowToMap, q)
 }
 

